numba/tests/test_typedlist.py

```diff
@@ -689,6 +826,319 @@ def foo():
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(list, type(foo_false_received))
+
+    def test_list_and_square_bracket_builtin_from_iter(self):
+        def foo():
+            l = list([1, 2, 3])
+            return l
+        foo_true, foo_false = self._njit_both(foo)
+        foo_true_received, foo_false_received = foo_true(), foo_false()
+        foo_true_expected, foo_false_expected = List(), list([1, 2, 3])
+        [foo_true_expected.append(i) for i in (1, 2, 3)]
+        self.assertEqual(foo_true_expected, foo_true_received)
+        self.assertEqual(foo_false_expected, foo_false_received)
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(list, type(foo_false_received))
+
+    @skip_py2
+    def test_dict_in_list_for_square_bracket_builtin(self):
+        def foo():
+            l = [{"a": 1}]
+            return l
+        d = Dict()
+        d["a"] = 1
+        foo_true, foo_false = self._njit_both(foo)
+        foo_true_received, foo_false_received = foo_true(), foo_false()
+        foo_true_expected, foo_false_expected = List(), [d]
+        foo_true_expected.append(d)
+        self.assertEqual(foo_true_expected, foo_true_received)
+        self.assertEqual(foo_false_expected, foo_false_received)
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(list, type(foo_false_received))
+
+    def test_square_bracket_builtin_from_nested_iter(self):
+        def foo():
+            l = [[1, 2, 3], [4, 5, 6]]
+            return l
+        foo_true, foo_false = self._njit_both(foo)
+        foo_true_received, foo_false_received = foo_true(), foo_false()
+        foo_true_expected, foo_false_expected = List(), list([[1, 2, 3],
+                                                              [4, 5, 6]])
+        a = List()
+        [a.append(i) for i in (1, 2, 3)]
+        b = List()
+        [b.append(i) for i in (4, 5, 6)]
+        foo_true_expected.append(a)
+        foo_true_expected.append(b)
+
+        self.assertEqual(foo_true_expected, foo_true_received)
+        self.assertEqual(foo_false_expected, foo_false_received)
+
+        self.assertEqual(List, type(foo_true_received))
+        self.assertEqual(List, type(foo_true_received[0]))
+        self.assertEqual(List, type(foo_true_received[1]))
+
+        self.assertEqual(list, type(foo_false_received))
+        self.assertEqual(list, type(foo_false_received[0]))
+        self.assertEqual(list, type(foo_false_received[1]))
+
+    def test_square_bracket_builtin_from_iter_type_coercion(self):
+        def foo():
+            l = [1, 1.0]
+            return l
+        foo_true, foo_false = self._njit_both(foo)
+        foo_true_expected = List()
+        foo_true_expected.append(1.0)
+        foo_true_expected.append(1.0)
+        foo_false_expected = [1.0, 1.0]
+        foo_true_received, foo_false_received = foo_true(), foo_false()
+        # FIXME: the first item is coerced to float
```

Comment:
This is also strange because `[1, 1.0]` would not be accepted as an argument to a function where `disable_reflected_list` is set.

Reply:
indeed.